src/main/java/org/checkerframework/specimin/MustImplementMethodsVisitor.java

```diff
@@ -0,0 +1,151 @@
+import com.github.javaparser.ast.visitor.ModifierVisitor;
+import com.github.javaparser.ast.visitor.Visitable;
+import com.github.javaparser.resolution.declarations.ResolvedMethodDeclaration;
+import com.github.javaparser.resolution.declarations.ResolvedParameterDeclaration;
+import java.util.Set;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * If a used class includes methods that must be implemented (because it extends an abstract class
+ * or implements an interface that requires them), this visitor marks them for preservation. Should
+ * run after the list of used classes is finalized.
+ */
+public class MustImplementMethodsVisitor extends ModifierVisitor<Void> {
+
+  /** Set containing the signatures of used member (fields and methods). */
+  private Set<String> usedMembers;
+
+  /** Set containing the signatures of used classes. */
+  private Set<String> usedClass;
+
+  /**
+   * Constructs a new SolveMethodOverridingVisitor with the provided sets of target methods, used
+   * members, and used classes.
+   *
+   * @param usedMembers Set containing the signatures of used members.
+   * @param usedClass Set containing the signatures of used classes.
+   */
+  public MustImplementMethodsVisitor(Set<String> usedMembers, Set<String> usedClass) {
+    this.usedMembers = usedMembers;
+    this.usedClass = usedClass;
+  }
+
+  /**
+   * Get the set containing the signatures of used members.
+   *
+   * @return The set containing the signatures of used members.
+   */
+  public Set<String> getUsedMembers() {
+    return usedMembers;
+  }
+
+  /**
+   * Get the set containing the signatures of used classes.
+   *
+   * @return The set containing the signatures of used classes.
+   */
+  public Set<String> getUsedClass() {
+    return usedClass;
+  }
+
+  @Override
+  @SuppressWarnings("nullness:return") // ok to return null, because this is a void visitor
+  public Visitable visit(ClassOrInterfaceDeclaration type, Void p) {
+    if (type.getFullyQualifiedName().isPresent()
+        && usedClass.contains(type.getFullyQualifiedName().get())) {
+      return super.visit(type, p);
+    } else {
+      // the effect of not calling super here is that only used classes
+      // will actually be visited by this class
+      return null;
+    }
+  }
+
+  @Override
+  public Visitable visit(MethodDeclaration method, Void p) {
+    ResolvedMethodDeclaration overridden = getOverriddenMethod(method);
+    // two cases: the method is a solvable override, and we can show that
+    // it is abstract (before the ||), or we can't solve the override but there
+    // is an @Override annotation. This relies on the use of @Override when
+    // implementing required methods from interfaces in the target code,
+    // but unfortunately I think it's the best that we can do here. (@Override
+    // is technically optional, but it is widely used.)
+    if (isAbstract(overridden) || (overridden == null && isOverride(method))) {
```

Comment:
I am not sure what will happen if there are unsolved symbols in these methods. Technically, the current version of `UnsolvedSymbolVisitor` only solves symbols belonged to or used by target methods. If we preserve these methods, we could have a non-compiled output due to the unsolved symbols that they introduce.

Reply:
> If we preserve these methods, we could have a non-compiled output due to the unsolved symbols that they introduce.

The only things preserved about these methods are their signatures, so the concern here is that the parameter or return types will be unsolved. I'm not sure if this can occur in practice: we'll only need to preserve a method like this when dealing with a library that 1) we cannot control, like java.util and 2) an abstract class or interface from that library is implemented by a used class. I think 1) especially is important: for most libraries, we would remove the interface if it wasn't required, or _remove the method from the interface_ if it wasn't used (because the interface would be a synthetic method under our control).

Regardless, this is an improvement on the current situation: at present, in this situation Specimin's output already doesn't compile (because the compiler complains about an unimplemented method).

Reply:
As a follow up to my comment above, this problem does appear to be what's causing cf-6030 to fail now: `MustImplementMethodVisitor` is throwing an `UnsolvedSymbolException` when trying to get a return type. I'm going to make this more conservative, and only preserve these methods if their return type and parameter types are solvable, for now.